src/System.Text.Utf8/src/System/Text/Utf8/Utf8String.cs

```diff
@@ -50,9 +49,32 @@ public Utf8String(byte[] utf8bytes, int index, int length)
             _length = length;
         }
 
+        // TODO: reevaluate implementation
         public Utf8String(IEnumerable<UnicodeCodePoint> codePoints)
         {
-            throw new NotImplementedException();
+            int len = GetUtf8LengthInBytes(codePoints);
+            byte[] utf8bytes = new byte[len];
+            unsafe
+            {
+                fixed (byte* utf8bytesPinned = utf8bytes)
+                {
+                    ByteSpan span = new ByteSpan(utf8bytesPinned, len);
+                    foreach (UnicodeCodePoint codePoint in codePoints)
+                    {
+                        int encodedBytes;
+                        if (!Utf8Encoder.TryEncodeCodePoint(codePoint, span, out encodedBytes))
+                        {
+                            throw new ArgumentException("Invalid code point", "codePoints");
+                        }
+                        span = span.Slice(encodedBytes);
```

Comment:
not sure how to write it differently without breaking down foreach into while with enumerator. Could you expand it?